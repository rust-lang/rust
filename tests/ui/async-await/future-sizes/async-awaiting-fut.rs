//@ compile-flags: -Z print-type-sizes --crate-type lib
//@ edition:2021
//@ build-pass
//@ ignore-pass

async fn wait() {}

async fn big_fut(arg: [u8; 1024]) {}

async fn calls_fut(fut: impl std::future::Future<Output = ()>) {
    loop {
        wait().await;
        if true {
            return fut.await;
        } else {
            wait().await;
        }
    }
}

pub async fn test() {
    let fut = big_fut([0u8; 1024]);
    calls_fut(fut).await;
}
