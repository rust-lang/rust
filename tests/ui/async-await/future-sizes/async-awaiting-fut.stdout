print-type-size type: `{async fn body of test()}`: 3078 bytes, alignment: 1 bytes
print-type-size     discriminant: 1 bytes
print-type-size     variant `Unresumed`: 0 bytes
print-type-size     variant `Suspend0`: 3077 bytes
print-type-size         local `.__awaitee`: 3077 bytes, type: {async fn body of calls_fut<{async fn body of big_fut()}>()}
print-type-size     variant `Returned`: 0 bytes
print-type-size     variant `Panicked`: 0 bytes
print-type-size type: `std::mem::ManuallyDrop<{async fn body of calls_fut<{async fn body of big_fut()}>()}>`: 3077 bytes, alignment: 1 bytes
print-type-size     field `.value`: 3077 bytes
print-type-size type: `std::mem::MaybeUninit<{async fn body of calls_fut<{async fn body of big_fut()}>()}>`: 3077 bytes, alignment: 1 bytes
print-type-size     variant `MaybeUninit`: 3077 bytes
print-type-size         field `.uninit`: 0 bytes
print-type-size         field `.value`: 3077 bytes
print-type-size type: `{async fn body of calls_fut<{async fn body of big_fut()}>()}`: 3077 bytes, alignment: 1 bytes
print-type-size     discriminant: 1 bytes
print-type-size     variant `Unresumed`: 1025 bytes
print-type-size         upvar `.fut`: 1025 bytes
print-type-size     variant `Suspend0`: 2052 bytes
print-type-size         upvar `.fut`: 1025 bytes
print-type-size         local `.fut`: 1025 bytes
print-type-size         local `..coroutine_field4`: 1 bytes, type: bool
print-type-size         local `.__awaitee`: 1 bytes, type: {async fn body of wait()}
print-type-size     variant `Suspend1`: 3076 bytes
print-type-size         upvar `.fut`: 1025 bytes
print-type-size         padding: 1025 bytes
print-type-size         local `..coroutine_field4`: 1 bytes, alignment: 1 bytes, type: bool
print-type-size         local `.__awaitee`: 1025 bytes, type: {async fn body of big_fut()}
print-type-size     variant `Suspend2`: 2052 bytes
print-type-size         upvar `.fut`: 1025 bytes
print-type-size         local `.fut`: 1025 bytes
print-type-size         local `..coroutine_field4`: 1 bytes, type: bool
print-type-size         local `.__awaitee`: 1 bytes, type: {async fn body of wait()}
print-type-size     variant `Returned`: 1025 bytes
print-type-size         upvar `.fut`: 1025 bytes
print-type-size     variant `Panicked`: 1025 bytes
print-type-size         upvar `.fut`: 1025 bytes
print-type-size type: `std::mem::ManuallyDrop<{async fn body of big_fut()}>`: 1025 bytes, alignment: 1 bytes
print-type-size     field `.value`: 1025 bytes
print-type-size type: `std::mem::MaybeUninit<{async fn body of big_fut()}>`: 1025 bytes, alignment: 1 bytes
print-type-size     variant `MaybeUninit`: 1025 bytes
print-type-size         field `.uninit`: 0 bytes
print-type-size         field `.value`: 1025 bytes
print-type-size type: `{async fn body of big_fut()}`: 1025 bytes, alignment: 1 bytes
print-type-size     discriminant: 1 bytes
print-type-size     variant `Unresumed`: 1024 bytes
print-type-size         upvar `.arg`: 1024 bytes
print-type-size     variant `Returned`: 1024 bytes
print-type-size         upvar `.arg`: 1024 bytes
print-type-size     variant `Panicked`: 1024 bytes
print-type-size         upvar `.arg`: 1024 bytes
print-type-size type: `std::mem::ManuallyDrop<bool>`: 1 bytes, alignment: 1 bytes
print-type-size     field `.value`: 1 bytes
print-type-size type: `std::mem::ManuallyDrop<{async fn body of wait()}>`: 1 bytes, alignment: 1 bytes
print-type-size     field `.value`: 1 bytes
print-type-size type: `std::mem::MaybeUninit<bool>`: 1 bytes, alignment: 1 bytes
print-type-size     variant `MaybeUninit`: 1 bytes
print-type-size         field `.uninit`: 0 bytes
print-type-size         field `.value`: 1 bytes
print-type-size type: `std::mem::MaybeUninit<{async fn body of wait()}>`: 1 bytes, alignment: 1 bytes
print-type-size     variant `MaybeUninit`: 1 bytes
print-type-size         field `.uninit`: 0 bytes
print-type-size         field `.value`: 1 bytes
print-type-size type: `std::task::Poll<()>`: 1 bytes, alignment: 1 bytes
print-type-size     discriminant: 1 bytes
print-type-size     variant `Ready`: 0 bytes
print-type-size         field `.0`: 0 bytes
print-type-size     variant `Pending`: 0 bytes
print-type-size type: `{async fn body of wait()}`: 1 bytes, alignment: 1 bytes
print-type-size     discriminant: 1 bytes
print-type-size     variant `Unresumed`: 0 bytes
print-type-size     variant `Returned`: 0 bytes
print-type-size     variant `Panicked`: 0 bytes
